File: python/CourseSchedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        
        graph = [[] for _ in range(numCourses)]
        visited = [0 for _ in range(numCourses)]
        taken = [0 for _ in range(numCourses)]
        
        for pair in prerequisites:
            course, pre = pair
            graph[course].append(pre)
                
        for i in range(numCourses):
            if not self.dfs(i, visited, taken, graph):
                return False
        
        return True
    
    def dfs(self, i, visited, taken, graph):
        if taken[i]:
            return True
        
        if visited[i]:
            return False
        
        visited[i] = 1
        
        for pre in graph[i]:
            if not self.dfs(pre, visited, taken, graph):
                return False
        
        visited[i] = 0
        taken[i] = 1
        
        return True

File: java/CourseSchedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> pre = new ArrayList<List<Integer>>();
        for (int i = 0; i < numCourses; i++){
            pre.add(new ArrayList<Integer>());
        }
        
        for (int i= 0; i < prerequisites.length; i++){
            pre.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }
        
        boolean[] visited = new boolean[numCourses];
        boolean[] taken = new boolean[numCourses];
        
        for (int i = 0; i < numCourses; i++){
            if (!dfs(i, visited, taken, pre)){
                return false;
            }
        }
        
        return true;
    }
    
    private boolean dfs(int courseId, boolean[] visited, boolean[] taken, List<List<Integer>> pre){
        if (taken[courseId]){
            return true;
        }
        
        if (visited[courseId]){
            return false;
        }
        
        visited[courseId] = true;
        for (int i : pre.get(courseId)){
            if (!dfs(i, visited, taken, pre)){
                return false;
            }
        }
        
        visited[courseId] = false;
        taken[courseId] = true;
        
        return true;
    }
}
